### ISSUE TITLE
BDS - Multi securties and fields

### ISSUE THREAD

**Issue by Carcajou01** [Rblp/Rblpapi#151]
Hello,

I'm new in R. I'm working on Mortgages market and I use BBG API through Excel VBA (see BBCOM_WRAPPER) to get my BBG data. I'm reviewing my process in order to switch to R and RBlpapi.

I'd like to request a list of bulk data for a list of fields and a list of securities. 

For example :
securities <- c("TDA 29 A2 MTGE" , "PARGN 24 A2 MTGE")
fields <- c("MTG_HIST_WAC","MTG_HIST_WAM")
bds(securities, fields)

error : more than 1 security selected

I guess we can't request more than 1 security and 1 field for this type of request. Nevertheless, is there a fast and clean way to do it instead of repeating bds querries ? :)

Many thanks and great work !
Paul

**Comment by eddelbuettel** [comment]
For what we know the API for `bds()` supports precisely one security.   So you answered your question already:

> repeating bds querries

I see no other way.

**Comment by csrvermaak** [comment]
I'm not 100% sure about this should be closed, it seems as if multiple securities are supported - the documentation for bds() states, for return Value:

> A list with as many entries as there are entries in
_securities_; each list contains a data.frame with one row
per observations and as many columns as entries in
_fields_. If the list is of length one, it is collapsed into
a single data frame.

**Comment by eddelbuettel** [comment]
Ok, good point, I'll reopen.  

There is clearly a mismatch between the documentation you cite (which maybe we just need to fix?) and the line in `bds()` which does

```r
if (length(securities)!=1L) stop("more than one security submitted.", call.=FALSE)
```

I just poked at calling the underlying code and called `bds_impl()` directly; it did not go well.  Maybe the documentation is out of whack.

**Comment by johnlaing** [comment]
I can see cases for a variety of different behaviors -- multiple securities, multiple fields, both... I'm not sure what my preference is. For now I think we should just update the documentation to match the current behavior, with the caveat that this may change in the future.

**Comment by eddelbuettel** [comment]
I want to poke back at the examples in the Bloomberg API tarball.  If it can support multiple securities we may as well.  If not, then we just adjust documentation as you suggest.

**Comment by johnlaing** [comment]
It can, it becomes a question of how we process them. Unlike `bdh` or other time-series-like calls, there isn't necessarily a natural way to combine the multiple results. We could, of course, just return a list. I'm not opposed to that, but I don't see that it adds much value over letting the user iterate or `lapply`. And it does require some work on the `bds_impl` internals.

**Comment by eddelbuettel** [comment]
I agree on 'simpler is better'. I just wonder what we'd gain. Looking at `src/bds.cpp` I do not see too much added complexity.

**Comment by armstrtw** [comment]
If bbg supports the call, then we just drop them in a list like we do w/ BDH.  If not, then let's change the doc.

**Comment by johnlaing** [comment]
OK, here's why this is hard. Bloomberg supports the call, but the way we're retrieving results doesn't play well with what they give back.

Ignore how we're processing the results for now, let's just look at what Bloomberg gives. An example with in which we pass 2 securities:
```R
> invisible(Rblpapi:::bds_Impl(con, c("CPI YOY Index", "CPI INDX Index"), "ECO_FUTURE_RELEASE_DATE_LIST", NULL, c("START_DT"="20160401", "END_DT"="20160430"), TRUE, NULL))
ReferenceDataResponse = {
    securityData[] = {
        securityData = {
            security = "CPI YOY Index"
            eidData[] = {
            }
            fieldExceptions[] = {
            }
            sequenceNumber = 0
            fieldData = {
                ECO_FUTURE_RELEASE_DATE_LIST[] = {
                    ECO_FUTURE_RELEASE_DATE_LIST = {
                        Release Dates & Times = "2016/04/14 08:30:00"
                    }
                }
            }
        }
    }
}
```
We only got one result back, that for the first security. Let's try another call with a new security:
```R
> invisible(Rblpapi:::bds_Impl(con, "CPI CHNG Index", "ECO_FUTURE_RELEASE_DATE_LIST", NULL, c("START_DT"="20160401", "END_DT"="20160430"), TRUE, NULL))
ReferenceDataResponse = {
    securityData[] = {
        securityData = {
            security = "CPI INDX Index"
            eidData[] = {
            }
            fieldExceptions[] = {
            }
            sequenceNumber = 1
            fieldData = {
                ECO_FUTURE_RELEASE_DATE_LIST[] = {
                    ECO_FUTURE_RELEASE_DATE_LIST = {
                        Release Dates & Times = "2016/04/14 08:30:00"
                    }
                }
            }
        }
    }
}
```
Oh look, there's the second result from our first call. Totally unrelated to the second call.

I've been bitten by this before in `Rbbg`. I'm sure it's possible to solve, but I question if it's worth it at the moment.

**Comment by eddelbuettel** [comment]
On the other hand:

```sh
~/blpapi_cpp_3.8.18.1$ ./SimpleRefDataExample_64                                                                                                           
SimpleRefDataExample
Sending Request: ReferenceDataRequest = { securities[] = { "IBM US Equity" "MSFT US Equity" } fields[] = { "PX_LAST" "DS002" } }
SessionConnectionUp = {
    server = "..........."
}

SessionStarted = {
}

ServiceOpened = {
    serviceName = "//blp/refdata"
}

ReferenceDataResponse = {
    securityData[] = {
        securityData = {
            security = "IBM US Equity"
            eidData[] = {
            }
            fieldExceptions[] = {
            }
            sequenceNumber = 0
            fieldData = {
                PX_LAST = 148.750000
                DS002 = "INTL BUSINESS MACHINES CORP"
            }
        }
        securityData = {
            security = "MSFT US Equity"
            eidData[] = {
            }
            fieldExceptions[] = {
            }
            sequenceNumber = 1
            fieldData = {
                PX_LAST = 53.760000
                DS002 = "MICROSOFT CORP"
            }
        }
    }
}

Press ENTER to quit

~/blpapi_cpp_3.8.18.1$ 
```

**Comment by eddelbuettel** [comment]
But I agree that a rewrite may not be worth it.

**Comment by johnlaing** [comment]
I think that example fits a `bdp` use case, which we support:
```R
> bdp(c("IBM US Equity", "MSFT US Equity"), c("PX_LAST", "DS002"))
               PX_LAST                       DS002
IBM US Equity   148.77 INTL BUSINESS MACHINES CORP
MSFT US Equity   53.87              MICROSOFT CORP
```
In that example both `securityData[]` chunks are contained in the same `ReferenceDataResponse`, which allows us to process correctly.

My vote is still very much in the "fix the documentation" camp.

**Comment by Carcajou01** [comment]
Hi,

Thanks for your comments ! 

As  mentionned, I used to work with VBA (Google : BBCOM_WRAPPER) for requesting Bloomberg API.
With that tool you can request several securities and fields for bulk data. Nevertheless you get back a tab of tab of tab.... which is not easy to use especially in VBA.

I'm new in R so I can't really help you whether you should "fix the doc" or not but I know that Matlab use "cells" data structure to fix the problem (http://fr.mathworks.com/help/datafeed/blp.getbulkdata.html).

Once again, thanks for your time !

**Comment by eddelbuettel** [comment]
tl;dr:  `help(sapply)` or `help(lapply)`

longer answer:   Most intermediate to advanced R tutorials will show you how to loop over a set and combine the results.  R is good at that, but it requires a bit of practice and reading.

**Comment by klin333** [comment]
Hi guys, a small code change I attempted was able to allow multiple securities per bds call. I think the speed up in total request time over multiple securities is almost 50 times faster than lapply on 1 security per bds call. 

Code patch attached. It is currently super buggy and causes R to crash sometimes, probably segfaults somewhere. Not surprised given how badly I am using Rcpp - I couldn't figure out how to append lists... Roughly over 500 securities and 500 bds fields per security, it takes 1.5 hours and crashes R roughly twice. I also suspect it drops a few results here and there - again not surprised given how badly I am using Rcpp to stitch together request return payloads. 

If interested, please help investigate the bugs in my patch and incorporate into rblpapi, thanks. If not, please delete this post as my buggy patch will cause bad R crashes and sporadically dropped results. 

[fast_bds.txt](https://github.com/Rblp/Rblpapi/files/3484907/fast_bds.txt)

**Comment by eddelbuettel** [comment]
Thanks, but better to make it proper pull request.
